hibernate-core/src/main/java/org/hibernate/dialect/Dialect.java

```diff
@@ -249,55 +264,181 @@ protected void registerDefaultColumnTypes() {
+				case VARBINARY:
+					registerColumnType( typeCode, maxVarBinaryLength, columnType(typeCode) );
+					registerColumnType( typeCode, columnType(LONGVARBINARY) );
+					break;
+				default:
+					registerColumnType( typeCode, columnType(typeCode) );
+			}
+		}
+	}
 
-		registerColumnType( Types.REAL, "real" );
-		registerColumnType( Types.FLOAT, "float($p)" );
-		registerColumnType( Types.DOUBLE, "double precision" );
+	/**
+	 * A list of JDBC types that we expect to be supported on all databases.
+	 */
+	private static final List<Integer> ANSI_SQL_TYPES = List.of(
+			BOOLEAN,
+			TINYINT, SMALLINT, INTEGER, BIGINT,
+			REAL, FLOAT, DOUBLE,
+			NUMERIC, DECIMAL,
+			DATE,
+			TIME, TIME_WITH_TIMEZONE,
+			TIMESTAMP, TIMESTAMP_WITH_TIMEZONE,
+			CHAR, VARCHAR, CLOB,
+			NCHAR, NVARCHAR, NCLOB,
+			BINARY, VARBINARY, BLOB
+	);
 
-		//these are pretty much synonyms, but are considered
-		//separate types by the ANSI spec, and in some dialects
-		registerColumnType( Types.NUMERIC, "numeric($p,$s)" );
-		registerColumnType( Types.DECIMAL, "decimal($p,$s)" );
+	/**
+	 * The JDBC type codes of types supported by this SQL dialect, from the lists
+	 * defined by {@link Types} and {@link SqlTypes}.
+	 * <p>
+	 * This method may be overridden by concrete {@code Dialect}s as an alternative
+	 * to calling {@link #registerColumnType(int, String)}. In this case,
+	 * {@link #columnType(int)} should also be overridden.
+	 * <p>
+	 * Note that {@link Types#LONGVARCHAR}, {@link Types#LONGNVARCHAR} and
+	 * {@link Types#LONGVARBINARY} are considered synonyms for their
+	 * non-{@code LONG} counterparts, and should not be included in the returned
+	 * array.
+	 *
+	 * @return an array of types from {@link SqlTypes}
+	 *
+	 * @see SqlTypes
+	 * @see #columnType(int)
+	 */
+	protected List<Integer> getSupportedJdbcTypeCodes() {
+		return ANSI_SQL_TYPES;
+	}
 
-		registerColumnType( Types.DATE, "date" );
-		registerColumnType( Types.TIME, "time" );
-		registerColumnType( Types.TIMESTAMP, "timestamp($p)" );
-		registerColumnType( Types.TIMESTAMP_WITH_TIMEZONE, "timestamp($p) with time zone" );
-		// type included here for completeness but note that
-		// very few databases support it, and the general
-		// advice is to caution against its use (for reasons,
-		// check the comments in the Postgres documentation).
-		registerColumnType( Types.TIME_WITH_TIMEZONE, "time with time zone" );
+	/**
+	 * The column type name for a given JDBC type code defined in {@link Types} or
+	 * {@link SqlTypes}. This default implementation returns the ANSI-standard type
+	 * name.
+	 * <p>
+	 * This method may be overridden by concrete {@code Dialect}s as an alternative
+	 * to calling {@link #registerColumnType(int,String)}.
+	 *
+	 * @param jdbcTypeCode a JDBC type code
+	 * @return a column type name, with $l, $p, $s placeholders for length, precision, scale
+	 *
+	 * @see SqlTypes
+	 * @see #getSupportedJdbcTypeCodes()
+	 */
+	protected String columnType(int jdbcTypeCode) {
```

Comment:
Maybe it's not worth doing as part of this PR, but I would personally love to see an actual contract here instead of returning a `String` e.g.

```
interface DdlType {
    String getType(Size size);
}
```

This way, we could handle some of the strange types that use bit lengths instead of bytes. Apart from that, such a contract would be extendible in various ways that makes future features easier possible. Something that would be awesome to add is something like:

```
default boolean needsCheckConstraintForSize() {
  return false;
} 
default String generateCheckConstraint(Size size) {
  return null;
}
```

which we could use to allow emitting check constraints for LONGVARCHAR/LONGVARBINARY. Maybe it's even worth passing the `Column` here to allow access to more metadata and enable more use cases?

Apart from that, a previous idea that came to mind was to capture somehow which JDBC API to use for a DDL type e.g. CLOB vs. String APIs to be able to use a clob DDL type even if LONGVARCHAR was requested, so that we can fit the requested size.

Reply:
Well I'm definitely not going to do anything additional in this PR, which is already more than big enough.